### ISSUE TITLE
Integration Tracking (WC Subscriptions): Ensure subscription orders use a supported payment method when using checkout block

### ISSUE THREAD

**Issue by haszari** [woocommerce/woocommerce-gutenberg-products-block#3252]
## Is your feature request related to a problem? Please describe.
Not all payment gateways [support WooCommerce Subscriptions](https://docs.woocommerce.com/document/subscriptions/enabling-payment-gateways-for-subscriptions/).

For example, on my store, in the shortcode checkout, onlyStripe and PayPal available when the order includes a subscription. 

<img width="1806" alt="Screen Shot 2020-10-08 at 1 39 03 PM" src="https://user-images.githubusercontent.com/4167300/95402632-b2740d00-096c-11eb-9106-8df3667c257d.png">

<img width="1809" alt="Screen Shot 2020-10-08 at 1 39 31 PM" src="https://user-images.githubusercontent.com/4167300/95402641-bbfd7500-096c-11eb-9130-e3b620fc1aee.png">

In checkout block, all the payment methods are shown.

<img width="1240" alt="Screen Shot 2020-10-08 at 1 47 39 PM" src="https://user-images.githubusercontent.com/4167300/95402696-d6375300-096c-11eb-9fad-7e6363a09cb7.png">

We'll need to add support for this in the checkout block payments components.

**Comment by senadir** [comment]
This is a confusing point that we need to clarify with @woocommerce/lambda.
Technically speaking, WooCommerce Subscription supports all payment methods, if you're creating a manual subscription. Only certain payment methods support automatic renewals, and only a few of them support advanced features like multiple subscriptions.

Subscriptions will alter your cart or show a notice if you have a conflicting cart, like having multiple subscriptions and only having PayPal standard.

In your photo, if you disable Stripe and Paypal, other payment methods will show up instead.

https://docs.woocommerce.com/document/subscriptions/payment-gateways/

**Comment by senadir** [comment]
@nerrad it seems subscriptions filters out certain payment methods from the checkout, we don't have a way to do that, I assumed the code should be compatible with our payment code.
Currently, Subscriptions filter out the unsupported payment methods via a filter called `woocommerce_available_payment_gateways`

That filter is part of `get_available_payment_gateways` which is used throughout the codebase to decide what payment method to show.

I'm considering adding `woocommerce_available_payment_gateways` to our payment gateway registration code.

**Comment by nerrad** [comment]
This is something that should ideally be filtered at the server level when the payment methods are being setup for the checkout.  I would start investigating if this is something that can be done [in this file](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/src/Payments/Api.php).

Now, the above applies only if subscriptions filters out available payment gateways given what is known to be in the cart/checkout. If it's possible that the checkout/cart contents might be changed on the fly (or switch between a subscription and a non-subscription type of item), then there likely will need to be additional filtering happen client side as well which could be more tricky and we'll have to have some discussion on that likely.

**Comment by senadir** [comment]
If we ignore express payment methods, then payment methods, in general, are not available in Cart (the only place where you could change the products), so by the time you reach Checkout, the products are final.

If we somehow allow editing products in Checkout (it feels like someday, we will add the ability to delete orders), then we will need to add that support directly in the plugin code, so in `canMakePayment`.

This means we will have to provide subscription support for each payment method we have, a bit of an undertaking.

**Comment by nerrad** [comment]
an i1 outcome initially looks like this can be handled server side from my perspective and would match the current behaviour (roughly) for the existing shortcode cart and checkout flow (from what's already been reported in this issue).

**Comment by haszari** [comment]
I've implemented a start on this in https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3500 - just firing the existing [core hook (`woocommerce_available_payment_gateways `)](https://github.com/woocommerce/woocommerce/blob/02cf0dfaed5923513de0c88add597d1560c2cfd2/includes/class-wc-payment-gateways.php#L160) for now.

This doesn't work yet, because of the way we handle dynamically-available payment methods. COD is the only example so far. COD is dependent on shipping. In JS in `usePaymentMethodRegistration` [we add them all back](https://github.com/woocommerce/woocommerce-gutenberg-products-block/blob/trunk/assets/js/base/context/cart-checkout/payment-methods/use-payment-method-registration.js#L163), i.e. essentially ignore the list of valid gateways from server.

I'm looking into how we can resolve this to combine the two needs:

- Payment methods like COD might not be included in "available" list (from server), though might dynamically become available (e.g. if shopper changes shipping).
- Other payment methods (like BACS, or COD!) should be excluded because they are not supported for the current items in the cart.

**Comment by haszari** [comment]
Update - did some more research to understand how this works, and if we should implement this in JS.

## TLDR
We might need to implement a new extensibility interface to control which payment methods are shown in checkout (dependent on order, e.g. cart contents or shipping method).

### How WC Subscriptions filters out payment methods
Payment methods declare what features they support via a [`supports` member and method](https://github.com/woocommerce/woocommerce/blob/master/includes/abstracts/abstract-wc-payment-gateway.php#L417) (which allows filtering). The features themselves are not defined anywhere that I can find; it looks like extensions can add new features and check for them.

For example, Stripe declares that it supports various subscriptions features here: 

https://github.com/woocommerce/woocommerce-gateway-stripe/blob/832aebe209ceeaf2b8f916dfec6ad5c0916bf3d1/includes/class-wc-gateway-stripe.php#L92-L108

To bring it back to what this issue needs - WC Subscriptions checks things like `supports( 'subscriptions' )` to determine which gateways to display.

### So .. should we (re)implement similar logic/API in JS?
We could expose the gateway `supports` array to js as script data; we already have a `supports` member, perhaps this was the original intention. 

Then we could adapt (port/fork) the the subscriptions code that filters the payment methods into JS. We'd probably want to do this generically, as a JS extensibility API.

Because we have COD dynamically available in checkout depending on shipping, I suspect we're going to need a new API somewhere (JS or PHP). 

cc @nerrad - FYI, I think this issue is more interesting than we hoped. Interested to hear if you have any ideas.

Next steps: I'll scope out what this JS api might look like, and any impact on the code related to COD availability. 

We'll also need to consider what other extensions might want to control payment method availability, and whether we want to expose such an API. If this is only needed by Subscriptions, maybe it's better to special-case it in checkout block code (i.e. checkout block has special code that fires if cart has subscriptions).

**Comment by senadir** [comment]
I think the suggestions you made here Rua are good, to give a summary of I have in mind:
- Make supports part of the registered payment methods, and somehow accessible.
- Provide some sort of global payment control, that can filter out methods, we might want to move ordering to it as well.
- That interface can be extended by third-party code (like subs) to introduce new filters.

We still need to hash how this might work without breaking all payment methods.

**Comment by nerrad** [comment]
Great research here Rua!

Ideally we'd implement this in a way that the `supports` feature is something that we simply need to implement as a part of the passed through data to the client so we end up with an object structure somewhat like this:

```js
const paymentMethodSupports = {
    stripe: [ 'products',
			'refunds',
			'tokenization',
			'add_payment_method',
			'subscriptions',
			'subscription_cancellation',
			'subscription_suspension',
			'subscription_reactivation',
			'subscription_amount_changes',
			'subscription_date_changes',
			'subscription_payment_method_change',
			'subscription_payment_method_change_customer',
			'subscription_payment_method_change_admin',
			'multiple_subscriptions',
			'pre-orders',
                        ],
    'paypal_standard': []
    /** ...etc **/
}
```
Then we can introduce a new filterable `order_type` key on the checkout endpoint response that would be an array of canonical "types" describing the type of order. Subscriptions would use this filter to:

- check the order data
- add its "types" (eg. `multiple_subscriptions`) to the response given the contents of the order.

Then js side, when `usePaymentMethodRegistration` is looping through the registered payment methods, it can check `order_type` from the checkout data and match against `paymentMethodSupports` data passed from the server for the given payment method to determine whether the payment method is actually registered or not.

This method would allow integration completely at the server level and subscriptions wouldn't need any client code for this specific case.

**Comment by haszari** [comment]
**Update** I've prototyped a lean approach in https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/3500. It's not quite what was proposed in above comments – those approaches didn't quite pan out initially (though we may come back to them).

@senadir and @nerrad (if you like) can you take a look at that PR and feedback? Most of the info is in the PR description, but the code works for the obvious use case. 

We might need to go wider for feedback here, or do some research to get more data about this need across other extensions. Feel free to ping in others for feedback, and/or I can P2 this for wider discussion next week.
